### PR TITLE
Add External Account Binding support.

### PR DESCRIPTION
**Pull request by robaman**
https://ietf-wg-acme.github.io/acme/draft-ietf-acme-acme.html#rfc.section.7.3.5

### CODE REVIEW

**Comment by robaman** [comment]
Thanks a lot for reviewing this! 
Will have a closer look at your comments when I get back from vacation in about 2 weeks.

File: providers/dns/dns_providers.go [hunk 1]
```diff
@@ -35,8 +34,6 @@ func NewDNSChallengeProviderByName(name string) (acme.ChallengeProvider, error)
 	var err error
 	var provider acme.ChallengeProvider
 	switch name {
-	case "azure":
```
[the remark this answers is not in the document]
Comment (robaman) [reply providers/dns/dns_providers.go]: Will fix.

File: cli_handlers.go [hunk 2]
```diff
@@ -250,7 +251,31 @@ func run(c *cli.Context) error {
 			logger().Fatal("You did not accept the TOS. Unable to proceed.")
 		}
 
-		reg, err := client.Register(accepted)
+		var reg *acme.RegistrationResource
+		var err error
+
+		kid := c.GlobalString("kid")
+		hmacEncoded := c.GlobalString("hmac")
```
[the remark this answers is not in the document]
Comment (robaman) [reply cli_handlers.go]: Moving them inside the if.

File: acmev2/client.go [hunk 3]
```diff
@@ -183,6 +184,57 @@ func (c *Client) Register(tosAgreed bool) (*RegistrationResource, error) {
 	return reg, nil
 }
 
+// Register the current account to the ACME server.
+func (c *Client) RegisterWithExternalAccountBinding(tosAgreed bool, kid string, hmac []byte) (*RegistrationResource, error) {
+	if c == nil || c.user == nil {
+		return nil, errors.New("acme: cannot register a nil client or user")
+	}
+	logf("[INFO] acme: Registering account (EAB) for %s", c.user.GetEmail())
```
[the remark this answers is not in the document]
Comment (robaman) [reply acmev2/client.go]: That's a good idea. Looking into it.

File: acmev2/jws.go [hunk 4]
```diff
@@ -87,6 +87,30 @@ func (j *jws) signContent(url string, content []byte) (*jose.JSONWebSignature, e
 	return signed, nil
 }
 
+func (j *jws) signEABContent(url, kid string, hmac, content []byte) (*jose.JSONWebSignature, error) {
+	signer, err := jose.NewSigner(
+		jose.SigningKey{Algorithm: jose.HS256, Key: hmac},
+		&jose.SignerOptions{
+			EmbedJWK:    false,
+			NonceSource: j,
```
[the remark this answers is not in the document]
Comment (robaman) [reply acmev2/jws.go]: From my understanding of https://ietf-wg-acme.github.io/acme/draft-ietf-acme-acme.html#rfc.section.7.3.5 there should be a nonce included with the /acme/new-account EAB http request. Am I misunderstanding you?

File: acmev2/client.go [hunk 5]
```diff
@@ -183,6 +184,57 @@ func (c *Client) Register(tosAgreed bool) (*RegistrationResource, error) {
...
+	accMsg := accountMessage{}
+	if c.user.GetEmail() != "" {
+		accMsg.Contact = []string{"mailto:" + c.user.GetEmail()}
+	} else {
+		accMsg.Contact = []string{}
+	}
+	accMsg.TermsOfServiceAgreed = tosAgreed
+
+
+	jwk := jose.JSONWebKey{Key: c.jws.privKey}
```
[the remark this answers is not in the document]
Comment (robaman) [reply acmev2/client.go]: Moved it.

**Comment by robaman** [comment]
Had to be a bit force-ful in my git pushing here. Hope it's ok. :) 
All the changes you requested but the EAB-requirement-feature is included in this second draft.

File: acmev2/jws.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (robaman) [reply acmev2/jws.go]: I understand you now. There shouldn't be a nonce in the inner EAB JWS.
Will fix this.

**Comment by ldez** [comment]
The CI fails, it's not the PR. I validated locally.

**Pull request merged**